File: 50dias/python/27.py
def LePalavrasDoArquivo():
    palavras = []
    input_file = open('input.txt', 'r')
    conteudo = input_file.readlines()
    input_file.close()

    for linha in conteudo:
        palavras.append(linha)
    return palavras

def RemoveColcheteEspacosDaString(entrada):
    entrada = entrada.replace(']', '')
    entrada = entrada.replace('[', '')
    entrada = entrada.replace(' ', '')
    return entrada

def ConverteListaStringsParaInteiros(entrada):
    entrada = RemoveColcheteEspacosDaString(entrada)

    inteirosEntrada = []
    if len(entrada) > 0:
        entrada = entrada.split(',')
        for palavra in entrada:
            inteirosEntrada.append(int(palavra))
    return inteirosEntrada

def ConverteArquivoParaListaInteiros():
    linhasDoArquivo = LePalavrasDoArquivo()

    listasDeInteiros = []
    for palavra in linhasDoArquivo:
        listaInteiros = ConverteListaStringsParaInteiros(palavra)
        listasDeInteiros.append(listaInteiros)
    return listasDeInteiros;

def OrdenaLista(lista):
    trocou = True
    while trocou:
        trocou = False
        for i in range(0, len(lista)-1):
            if lista[i] > lista[i+1]:
                aux = lista[i]
                lista[i] = lista[i+1]
                lista[i+1] = aux
                trocou = True

def OrdenaListas(listasLidas):
    for lista in listasLidas:
        OrdenaLista(lista)

'''
    Para encontrar o elemento faltante, você pode ordenar a 
    lista, visto que os elementos nao se repetem.
    exemplo:
        [4, 3, 1, 5, 9, 7, 8, 6, 10]
    ordenada:
        [1, 3, 4, 5, 6, 7, 8, 9, 10]
    O número 1 encontra-se na primeira posição
    mas o número 2 nao se encontra na segunda posição.
    Por este motivo o número DOIS está faltando.
    Se você encontrar todos elementos na posicao certa
    significa que o ultimo elemento é o faltante.
    Lembrando que os numeros (no exemplo) sao de 1 a 10
    e a lista tem 9 posicoes. Entao o numero DEZ
    está faltando.
'''

def EncontraElementoFaltante(lista):
    tamanhoLista = len(lista)
    print(lista)
    for i in range(0, tamanhoLista):
        if lista[i] != i + 1:
            return i+1
    return tamanhoLista+1

def EncontraElementoFaltanteNasListas(listasLidas):
    for lista in listasLidas:
        elemento = EncontraElementoFaltante(lista)
        print('O elemento faltando: {}'.format(elemento))

def main():
    listasLidas = ConverteArquivoParaListaInteiros()
    OrdenaListas(listasLidas)
    EncontraElementoFaltanteNasListas(listasLidas)

main()

File: 50dias/cpp/27.cpp
#include <iostream>
#include <fstream>
#include <list>
#include <sstream>

using namespace std;

list<string> LeArquivoTexto() {
    list<string> linhasDoArquivo;
    ifstream inputFile;
    inputFile.open("input.txt");

    string linha;
    while (getline(inputFile, linha))
        linhasDoArquivo.push_back(linha);
    inputFile.close();

    return linhasDoArquivo;
}

void RemoveColchetesEspacosDaString(string& entrada) {
    erase(entrada, '[');
    erase(entrada, ']');
    erase(entrada, ' ');
}

list<int> ConverteListaStringsParaInteiros(string entrada) {
    RemoveColchetesEspacosDaString(entrada);

    list<int> inteirosEntrada;
    if (entrada.size() > 0) {
        stringstream sstr(entrada);
        while (sstr.good()) {
            string substring;
            getline(sstr, substring, ',');
            inteirosEntrada.push_back(stoi(substring));
        }
    }

    return inteirosEntrada;
}

list<list<int>> ConverteArquivoParaListasInteiros() {
    list<string> linhasDoArquivo = LeArquivoTexto();

    list<list<int>> listasDeInteiros;
    for (string palavra : linhasDoArquivo) {
        list<int> listaInteiros = ConverteListaStringsParaInteiros(palavra);
        listasDeInteiros.push_back(listaInteiros);
    }

    return listasDeInteiros;
}

list<int> OrdenaLista(list<int>& lista) {
    bool trocou;
    list<int> listaOrdenada = lista;
    do {
        trocou = false;
        auto j = listaOrdenada.begin();
        auto i = j++;
        auto end = listaOrdenada.end();
        while (j != end) {
            if (*i > *j) {
                std::swap(*i, *j);
                trocou = true;
            }
            i++; j++;
        }
    } while (trocou);
    return listaOrdenada;
}

list<list<int>> OrdenaListas(list<list<int>> listasLidas) {
    list<list<int>> listasOrdenadas;
    for (auto lista : listasLidas) {
        list<int> listaOrdenada = OrdenaLista(lista);
        listasOrdenadas.push_back(listaOrdenada);
    }
    return listasOrdenadas;
}

/*
    Para encontrar o elemento faltante, voce pode ordenar
    a lista, visto que os elementos nao repetem
    exemplo:
        [4, 3, 1, 5, 9, 7, 8, 6, 10]
    ordenada:
        [1, 3, 4, 5, 6, 7, 8, 9, 10]
    O 1 está na primeira posicao.
    Porem o 2 nao está na segunda posicao.
    Entao o numero faltante é o 2.
    Se a lista inteira for percorrida e voce encontrar todos
    significa que o ultimo elemento (no caso do exemplo, o 10)
    é o elemento faltante, pois a lista tem todos os
    elementos de 1 a 9 nas posicoes corretas.
*/

int EncontraElementoFaltante(list<int> lista) {
    int tamanhoLista = lista.size();
    int posicao = 1;
    for (int elemento : lista) {
        if (elemento != posicao)
            return posicao;
        posicao++;
    }
    return tamanhoLista + 1;
}

void EncontraElementoFaltanteNasListas(list<list<int>> listasLidas) {
    for (auto lista : listasLidas) {
        int elemento = EncontraElementoFaltante(lista);
        cout << "Está faltando o numero " << elemento << endl;
    }
}

int main()
{
    list<list<int>> listasLidas = ConverteArquivoParaListasInteiros();
    list<list<int>> listasOrdenadas = OrdenaListas(listasLidas);
    EncontraElementoFaltanteNasListas(listasOrdenadas);

    return 0;
}

















